crates/solver/src/liquidity/zeroex.rs

```diff
@@ -9,39 +9,287 @@ use {
         liquidity_collector::LiquidityCollecting,
         settlement::SettlementEncoder,
     },
-    anyhow::Result,
+    anyhow::{anyhow, Context, Result},
     contracts::{GPv2Settlement, IZeroEx},
+    ethcontract::Address,
+    futures::{SinkExt, StreamExt},
     model::{order::OrderKind, TokenPair},
     primitive_types::{H160, U256},
+    reqwest::{IntoUrl, Url},
     shared::{
         ethrpc::Web3,
         http_solver::model::TokenAmount,
         recent_block_cache::Block,
-        zeroex_api::{Order, OrderRecord, OrdersQuery, ZeroExApi},
+        zeroex_api::{
+            websocket::{OrderRecord, OrderState, OrdersResponse},
+            Order,
+            OrdersQuery,
+            ZeroExApi,
+        },
     },
     std::{
         collections::{HashMap, HashSet},
         sync::Arc,
+        time::Duration,
     },
+    tokio::{net::TcpStream, sync::Mutex},
+    tokio_tungstenite::{tungstenite::Message, MaybeTlsStream, WebSocketStream},
+    tracing::Instrument,
 };
 
+const DEFAULT_ZEROEX_WEBSOCKET_API: &str = "wss://api.0x.org/orderbook/v1";
+
+type Socket = WebSocketStream<MaybeTlsStream<TcpStream>>;
+
+#[derive(Clone)]
+pub struct ZeroExWebsocketApi {
+    url: Url,
+}
+
+impl ZeroExWebsocketApi {
+    pub fn new(url: impl IntoUrl) -> Self {
+        Self {
+            url: url
+                .into_url()
+                .unwrap_or(Url::parse(DEFAULT_ZEROEX_WEBSOCKET_API).unwrap()),
```

Comment:
If parsing the URL we are passing in fails, I don't think we should silently fail to some default URL but instead either error or panic.